php/libraries/User.class.inc

```diff
@@ -65,6 +62,22 @@ class User extends UserPermissions
 
         $row = $DB->pselectRow($query, array('UID' => $username));
 
+        if (empty($row)) {
```

Comment:
@driusan review carefully

do you foresee issues with logic ?

Reply:
just the usual "0" issues with empty()..

Reply:
@driusan I meant with moving around the ordering of the calls in the user constructor. if you expand  the code u'll notice that I put this check and moved down a few variable declaration. Is any page relying on fetching an empty user ? 

(ie probably create_user)

Reply:
I thought either you or @xlecours had found one when you looked into it a few weeks ago or another reason, but I can't remember the details? (You're likely right about create_user.) If there are any, hopefully the intergration tests would catch it..

As far as changing the logic goes the only issue I see is that it doesn't really accomplish what the comment says it's trying to, because the early return makes it vulnerable to timing attacks, so a sufficiently advanced hacker could still derive the existence/non-existence of the username.

Reply:
yeah.. i'll put a little sleep function on it to confuse the Russian/Chinese hackers trying to break into loris...

this kind of avoids like 30 notices/warnings so if there is not functional issue I still like it there

Reply:
A `sleep` wouldn't prevent timing attacks, you need to ensure the same amount of work goes into processing both authenticated and unauthenticated requests to do so effectively.

Reply:
Why the mention of timing attacks here? According to the test plan you should update the front-end to say e.g. "A password reset email has been sent to the address corresponding to the username" right after the request is made -- regardless of whether that user exists. In fact that should be done before any of the server-side work so there's nothing to time from the perspective of the hacker.

Reply:
@ridz1208 still a good check to prevent flooding the logs tho

Reply:
also phishing =/= brute forcing =/= timing attacks but now I'm just being mean

Reply:
The comment says it's not returning an error so that phishers won't know that the user doesn't exist. They can still figure out whether the user exists by timing the response time with this implementation. Just because it's not an advanced technique for statistically deriving a password doesn't mean it's not a timing attack.

Reply:
alright, I just wanna clear up a couple things.

- whether the user exists or not, the "thank you ..." message is still being displayed
- if i have to keep the same amount of work done whether the user exists or not, then we have to figure out a way to do all that work without generating errors and warnings
- there is a typo on line 68 where it says "does" instead of "does not" that I just corrected

@johnsaigle, in this case, wouldnt you use a bruteforce technique in order to accomplish a phishing attack? <y terms might be confused but what I mean is:
Assuming the page returns a message like "this user does not exists" (which it doesnt), someone could write a script that accesses the forgot password page and try out usernames in a **brute force** manner (intelligent or not all combinations of names/words/wtv) to **get information** about existing users ?? am I wrong??

Reply:
Yeah all of that sounds right in terms of enumerating passwords 

> the page returns a message like "this user does not exists" (which it doesnt)

^ this is what I was trying to emphasize. As long as the page doesn't present a different response based on whether a user exists, and this is done client-side, there's no response to time. 

I might be mistaking something because I didn't read all the code for that patch but I was basing my comments on `modules/login/test/password_reset.md`. Everything looks good there.

Reply:
@johnsaigle yeah, I saw that yesterday in the test plan and changed it. if you look at the testplan I updated yesterday it should be corrected now :)